### PR TITLE
Dev make cleanup

### LINKED ISSUE #2839: Tidying up the make hierarchy
This is a component of #2816 and relates to #2811 

The current make hierarchy is a sloppy mess which I want to clean up but staying within the current make framework.  If we look at a typical compile line generated in the make

Type | Option
----|--------
Compiler Warnings | `-Wimplicit` `-Wpointer-arith` `-Wundef` `-Werror`
Optimisation stuff | `-O2` `-ffunction-sections` `-fno-jump-tables` `-fdata-sections` `-fno-inline-functions` `-mlongcalls` `-mtext-section-literals`
Language Spec | `-std=gnu11`
Linker stuff | `-Wl,-EL` `-nostdlib`
Preprocesssor defines | `` `__ets__` `EBUF_LWIP` `ICACHE_FLASH` `LWIP_OPEN_SRC` `MBEDTLS_USER_CONFIG_FILE="user_mbedtls.h"` `` `MEM_DEFAULT_USE_DRAM` `PBUF_RSV_FOR_WLAN` `USE_OPTIMIZE_PRINTF` ``

Type | Include List (not nec in this order)
----|--------
Dirs don't exist / contain headers | `<root>/` `<root>/include` `../smart` `../../include/eagle` 
Defaults | `./` `include` `../include``../../include`| `../spiffs` `../fatfs`
Core stuff | `../libc` `../lua` `../platform` 
Why expose? | `../spiffs` `../fatfs`
Packages | `../coap ` `../dht` `../http` `../mqtt` `../pcm` `../pm` `../sjson` `../sqlite3` `../u8g2lib/u8g2/src/clib` `../ucglib/ucg/src/clib` `../websocket`

This all need cleaning up:
-  The warnings and optimisation stuff are fine, except that we will be moving directories to `-Wall` `-Wextra` over time. 
-  Defines should be moved into a config header file or down into the sub make that needs them
-  Pathing in directories that don't exist or don't validly contain includes should be removed
-  We have some standard roots: `ROOTDIR` `APPDIR` `LUADIR` `INCDIR`
-  Most of the package includes are used in very specific instances so there is no point in including `app/http` (for example) in the search list; the one reference in  `app/http/httpclient.c` should simply refer to `#include "http/httpclient.h"` instead.

This is all pretty mechanistic tidying.  I just don't want any hard wired reference to `app` or `app/lua` in any Makefiles.

### PR DESCRIPTION
**Pull request by TerryE**
Fixes #2839

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.

This change doesn't affect the compiled binary, just the make process itself.  The best way to see this is with a `make V=1` before and after.

### CODE REVIEW

**Comment by nwf** [comment]
These changes all seem sensible.  While you're at it, can the `** HEALTH WARNING **` section of the main `Makefile` get tidied, too?  (Perhaps that's a bigger thing and should wait for a later PR.)

**Comment by TerryE** [comment]
OK I've now found why Travis failed.  The build process specifies EXTRA_CCFLAGS as a commandline parameter and doing this overrides all EXTRA_CCFLAGS setting inside the make so I need to make some adjustments to accommodate this.

**Comment by TerryE** [comment]
It might be good to get #2841 and this in before we do the master drop, but I think that they both touch some common files, so whoever goes second will need to do some fixups or rebaseline.  We should let #2841 go first and soon, I think.

File: app/Makefile [hunk 1]
```diff
@@ -54,35 +52,36 @@ endif # } PDIR
...
+	pm/libpm.a					\
+	esp-gdbstub/libgdbstub.a	\
+	net/libnodemcu_net.a		\
+	mbedtls/libmbedtls.a    	\
+	modules/libmodules.a		\
 	smart/smart.a 				\
 	uzlib/libuzlib.a 			\
 	$(OPT_SEL_COMPONENTS)
-
+# Ensure that the Espresif SDK is searched before the tool-chain's SDK (if any)
```
Comment (jmattsson) [inline app/Makefile]: I don't think this comment is accurate here. Copy paste slip?

File: app/user/Makefile [hunk 2]
```diff
@@ -38,13 +38,7 @@ DEFINES += -DESP_INIT_DATA_DEFAULT="\"$(SDK_DIR)/bin/esp_init_data_default_v05.b
...
 #
 
-INCLUDES := $(INCLUDES) -I $(PDIR)include
-INCLUDES += -I ./
-INCLUDES += -I ../../include/ets
-INCLUDES += -I ../libc
-INCLUDES += -I ../platform
-INCLUDES += -I ../lua
-INCLUDES += -I ../spiffs
+INCLUDES += -I ../../include/ets -I ../spiffs
```
Comment (jmattsson) [inline app/user/Makefile]: I don't think we have an `include/ets` either, we should be able to evict that too.

File: sdk-overrides/include/c_types.h [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#define ICACHE_FLASH
+#include_next "c_types.h"
```
Comment (jmattsson) [inline sdk-overrides/include/c_types.h]: Objection, your honour! :)
When I hit that other shiny green button just before, our `c_types.h` got decimated and should stay that way.

File: app/user/Makefile (hunk 2 above)
Comment (TerryE) [reply app/user/Makefile]: Will do

File: sdk-overrides/include/c_types.h (hunk 3 above)
Comment (TerryE) [reply sdk-overrides/include/c_types.h]: Yup, agreed. All part of the rebaseline TBD today. The reason for the wrapper is that ICACHE_FLASH was only used in this include.

File: app/Makefile (hunk 1 above)
Comment (TerryE) [reply app/Makefile]: I will check and remove

**Comment by TerryE** [comment]
I will do the rebaseline my late PM. I've got other commitments before then.

**Comment by TerryE** [comment]
I am note sure of the procedural niceties here. @nwf has reviewed this but not used the github review function do to this.  @jmattsson has also reviewed this and I've addressed these comments but not I am not sure how to mark this off as completed.  Anyway, if this Travis make completes then this is a verification of the make options as reviewed by two other committers, so I will then merge and squash this.

**Pull request merged**